csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/CompilerDiagnostic.cs

```diff
@@ -0,0 +1,45 @@
+using System.Collections.Concurrent;
+using System.IO;
+using Semmle.Util;
+
+namespace Semmle.Extraction.CSharp.Entities
+{
+    internal class CompilerDiagnostic : FreshEntity
+    {
+        private static readonly int limit = EnvironmentVariables.TryGetExtractorOption<int>("COMPILER_DIAGNOSTIC_LIMIT") ?? 1000;
+        private static readonly ConcurrentDictionary<string, int> messageCounts = new();
+
+        private readonly Microsoft.CodeAnalysis.Diagnostic diagnostic;
+        private readonly Compilation compilation;
+        private readonly int index;
+
+        public CompilerDiagnostic(Context cx, Microsoft.CodeAnalysis.Diagnostic diag, Compilation compilation, int index) : base(cx)
+        {
+            diagnostic = diag;
+            this.compilation = compilation;
+            this.index = index;
+            TryPopulate();
+        }
+
+        protected override void Populate(TextWriter trapFile)
+        {
+            // The below doesn't limit the extractor messages to the exact limit, but it's good enough.
+            var key = diagnostic.Id;
+            var messageCount = messageCounts.AddOrUpdate(key, 1, (_, c) => c + 1);
+            if (messageCount > limit)
+            {
+                if (messageCount == limit + 1)
+                {
+                    Context.Extractor.Logger.LogWarning($"Stopped logging {key} compiler diagnostics after reaching {limit}");
```

Comment:
@yo-h brought up today that we could log the total number of compiler errors for telemetry purposes.

Reply:
This is done in https://github.com/github/codeql/pull/15957/commits/205d6a3bc5e55e9a2e4cfa9e522cae1219c22461.